// honkit/lib/models/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const is_1 = __importDefault(require("is"));
const immutable_1 = __importDefault(require("immutable"));
const file_1 = __importDefault(require("./file"));
const pluginDependency_1 = __importDefault(require("./pluginDependency"));
const configDefault_1 = __importDefault(require("../constants/configDefault"));
const reducedObject_1 = __importDefault(require("../utils/reducedObject"));
const Config = immutable_1.default.Record({
    file: file_1.default(),
    values: configDefault_1.default,
}, "Config");
Config.prototype.getFile = function () {
    return this.get("file");
};
Config.prototype.getValues = function () {
    return this.get("values");
};
/**
 * Return minimum version of configuration,
 * Basically it returns the current config minus the default one
 * @return {Map}
 */
Config.prototype.toReducedVersion = function () {
    return reducedObject_1.default(configDefault_1.default, this.getValues());
};
/**
 * Render config as text
 * @return {Promise<String>}
 */
Config.prototype.toText = function () {
    return JSON.stringify(this.toReducedVersion().toJS(), null, 4);
};
/**
 * Change the file for the configuration
 * @param {File} file
 * @return {Config}
 */
Config.prototype.setFile = function (file) {
    return this.set("file", file);
};
/**
 * Return a configuration value by its key path
 * @param {String} key
 * @return {Mixed}
 */
Config.prototype.getValue = function (keyPath, def) {
    const values = this.getValues();
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'keyToKeyPath' does not exist on type 'Cl... Remove this comment to see the full error message
    keyPath = Config.keyToKeyPath(keyPath);
    if (!values.hasIn(keyPath)) {
        return immutable_1.default.fromJS(def);
    }
    return values.getIn(keyPath);
};
/**
 * Update a configuration value
 * @param {String} key
 * @param {Mixed} value
 * @return {Config}
 */
Config.prototype.setValue = function (keyPath, value) {
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'keyToKeyPath' does not exist on type 'Cl... Remove this comment to see the full error message
    keyPath = Config.keyToKeyPath(keyPath);
    value = immutable_1.default.fromJS(value);
    let values = this.getValues();
    values = values.setIn(keyPath, value);
    return this.set("values", values);
};
/**
 * Return a list of plugin dependencies
 * @return {List<PluginDependency>}
 */
Config.prototype.getPluginDependencies = function () {
    const plugins = this.getValue("plugins");
    if (is_1.default.string(plugins)) {
        // @ts-expect-error ts-migrate(2339) FIXME: Property 'listFromString' does not exist on type '... Remove this comment to see the full error message
        return pluginDependency_1.default.listFromString(plugins);
    }
    else {
        // @ts-expect-error ts-migrate(2339) FIXME: Property 'listFromArray' does not exist on type 'C... Remove this comment to see the full error message
        return pluginDependency_1.default.listFromArray(plugins);
    }
};
/**
 * Return a plugin dependency by its name
 * @param {String} name
 * @return {PluginDependency}
 */
Config.prototype.getPluginDependency = function (name) {
    const plugins = this.getPluginDependencies();
    return plugins.find((dep) => {
        return dep.getName() === name;
    });
};
/**
 * Update the list of plugins dependencies
 * @param {List<PluginDependency>}
 * @return {Config}
 */
Config.prototype.setPluginDependencies = function (deps) {
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'listToArray' does not exist on type 'Cla... Remove this comment to see the full error message
    const plugins = pluginDependency_1.default.listToArray(deps);
    return this.setValue("plugins", plugins);
};
/**
 * Update values for an existing configuration
 * @param {Object} values
 * @returns {Config}
 */
Config.prototype.updateValues = function (values) {
    values = immutable_1.default.fromJS(values);
    return this.set("values", values);
};
/**
 * Update values for an existing configuration
 * @param {Config} config
 * @param {Object} values
 * @returns {Config}
 */
Config.prototype.mergeValues = function (values) {
    let currentValues = this.getValues();
    values = immutable_1.default.fromJS(values);
    currentValues = currentValues.mergeDeep(values);
    return this.set("values", currentValues);
};
/**
 * Create a new config for a file
 * @param {File} file
 * @param {Object} values
 * @returns {Config}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'create' does not exist on type 'Class'.
Config.create = function (file, values) {
    return new Config({
        file: file,
        values: immutable_1.default.fromJS(values),
    });
};
/**
 * Create a new config
 * @param {Object} values
 * @returns {Config}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'createWithValues' does not exist on type... Remove this comment to see the full error message
Config.createWithValues = function (values) {
    return new Config({
        values: immutable_1.default.fromJS(values),
    });
};
/**
 * Convert a keyPath to an array of keys
 * @param {String|Array}
 * @return {Array}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'keyToKeyPath' does not exist on type 'Cl... Remove this comment to see the full error message
Config.keyToKeyPath = function (keyPath) {
    if (is_1.default.string(keyPath))
        keyPath = keyPath.split(".");
    return keyPath;
};
exports.default = Config;
