// honkit/lib/models/page.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const immutable_1 = __importDefault(require("immutable"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const file_1 = __importDefault(require("./file"));
const hash_1 = require("./hash");
const Page = immutable_1.default.Record({
    file: file_1.default(),
    // Attributes extracted from the YAML header
    attributes: immutable_1.default.Map(),
    // Content of the page
    content: String(),
    // Direction of the text
    dir: String("ltr"),
});
Page.prototype.getFile = function () {
    return this.get("file");
};
Page.prototype.getAttributes = function () {
    return this.get("attributes");
};
Page.prototype.getContent = function () {
    return this.get("content");
};
Page.prototype.getDir = function () {
    return this.get("dir");
};
/**
 * Return page as text
 * @return {String}
 */
Page.prototype.toText = function () {
    const attrs = this.getAttributes();
    const content = this.getContent();
    if (attrs.size === 0) {
        return content;
    }
    const frontMatter = `---\n${js_yaml_1.default.safeDump(attrs.toJS(), { skipInvalid: true })}---\n\n`;
    return frontMatter + (content || "");
};
/**
 * Return path of the page
 * @return {String}
 */
Page.prototype.getPath = function () {
    return this.getFile().getPath();
};
/**
 * Create a page for a file
 * @param {File} file
 * @return {Page}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'createForFile' does not exist on type 'C... Remove this comment to see the full error message
Page.createForFile = function (file) {
    return new Page({
        file: file,
    });
};
/**
 * Load a page for a file
 * @param {File} file
 * @param {string} content
 * @return {Page}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'loadFile' does not exist on type 'Class'... Remove this comment to see the full error message
Page.loadFile = function (file, content) {
    return new Page({
        file: file,
        content: content,
    });
};
// @ts-expect-error ts-migrate(2339) FIXME: Property 'fromJSON' does not exist on type 'Class'... Remove this comment to see the full error message
Page.fromJSON = function (json) {
    return new Page({
        file: new file_1.default(json.file),
        // Attributes extracted from the YAML header
        attributes: immutable_1.default.Map(json.atributes),
        // Content of the page
        content: json.content,
        // Direction of the text
        dir: json.dir,
    });
};
// @ts-expect-error ts-migrate(2339) FIXME: Property 'toJSON' does not exist on type 'Class'.
Page.toJSON = function (page) {
    return page.toJS();
};
Page.prototype.hash = function () {
    return hash_1.hashString(JSON.stringify(this.toJS()));
};
exports.default = Page;
