// honkit/lib/models/readme.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const immutable_1 = __importDefault(require("immutable"));
const file_1 = __importDefault(require("./file"));
const Readme = immutable_1.default.Record({
    file: file_1.default(),
    title: String(),
    description: String(),
});
Readme.prototype.getFile = function () {
    return this.get("file");
};
Readme.prototype.getTitle = function () {
    return this.get("title");
};
Readme.prototype.getDescription = function () {
    return this.get("description");
};
/**
 Create a new readme

 @param {File} file
 @param {Object} def
 @return {Readme}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'create' does not exist on type 'Class'.
Readme.create = function (file, def) {
    def = def || {};
    return new Readme({
        file: file,
        title: def.title || "",
        description: def.description || "",
    });
};
exports.default = Readme;
