// honkit/lib/models/summaryPart.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const immutable_1 = __importDefault(require("immutable"));
const summaryArticle_1 = __importDefault(require("./summaryArticle"));
/*
    A part represents a section in the Summary / table of Contents
*/
const SummaryPart = immutable_1.default.Record({
    level: String(),
    title: String(),
    articles: immutable_1.default.List(),
});
SummaryPart.prototype.getLevel = function () {
    return this.get("level");
};
SummaryPart.prototype.getTitle = function () {
    return this.get("title");
};
SummaryPart.prototype.getArticles = function () {
    return this.get("articles");
};
/**
 * Create a new level for a new child article
 *
 * @return {String}
 */
SummaryPart.prototype.createChildLevel = function () {
    const level = this.getLevel();
    const subArticles = this.getArticles();
    const childLevel = `${level}.${subArticles.size + 1}`;
    return childLevel;
};
/**
 * Create a SummaryPart
 *
 * @param {Object} def
 * @return {SummaryPart}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'create' does not exist on type 'Class'.
SummaryPart.create = function (def, level) {
    const articles = (def.articles || []).map((article, i) => {
        if (article instanceof summaryArticle_1.default) {
            return article;
        }
        // @ts-expect-error ts-migrate(2339) FIXME: Property 'create' does not exist on type 'Class'.
        return summaryArticle_1.default.create(article, [level, i + 1].join("."));
    });
    return new SummaryPart({
        level: String(level),
        title: def.title,
        articles: immutable_1.default.List(articles),
    });
};
exports.default = SummaryPart;
