// honkit/lib/models/templateEngine.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const nunjucks_1 = __importDefault(require("nunjucks"));
const immutable_1 = __importDefault(require("immutable"));
const TemplateEngine = immutable_1.default.Record({
    // Map of {TemplateBlock}
    blocks: immutable_1.default.Map(),
    // Map of Extension
    extensions: immutable_1.default.Map(),
    // Map of filters: {String} name -> {Function} fn
    filters: immutable_1.default.Map(),
    // Map of globals: {String} name -> {Mixed}
    globals: immutable_1.default.Map(),
    // Context for filters / blocks
    context: Object(),
    // Nunjucks loader
    loader: new nunjucks_1.default.FileSystemLoader("views"),
}, "TemplateEngine");
TemplateEngine.prototype.getBlocks = function () {
    return this.get("blocks");
};
TemplateEngine.prototype.getGlobals = function () {
    return this.get("globals");
};
TemplateEngine.prototype.getFilters = function () {
    return this.get("filters");
};
TemplateEngine.prototype.getShortcuts = function () {
    return this.get("shortcuts");
};
TemplateEngine.prototype.getLoader = function () {
    return this.get("loader");
};
TemplateEngine.prototype.getContext = function () {
    return this.get("context");
};
TemplateEngine.prototype.getExtensions = function () {
    return this.get("extensions");
};
/**
 Return a block by its name (or undefined)

 @param {String} name
 @return {TemplateBlock}
 */
TemplateEngine.prototype.getBlock = function (name) {
    const blocks = this.getBlocks();
    return blocks.find((block) => {
        return block.getName() === name;
    });
};
/**
 Return a nunjucks environment from this configuration

 @return {Nunjucks.Environment}
 */
TemplateEngine.prototype.toNunjucks = function (blocksOutput) {
    const loader = this.getLoader();
    const blocks = this.getBlocks();
    const filters = this.getFilters();
    const globals = this.getGlobals();
    const extensions = this.getExtensions();
    const context = this.getContext();
    const env = new nunjucks_1.default.Environment(loader, {
        // Escaping is done after by the asciidoc/markdown parser
        autoescape: false,
        // Syntax
        tags: {
            blockStart: "{%",
            blockEnd: "%}",
            variableStart: "{{",
            variableEnd: "}}",
            commentStart: "{###",
            commentEnd: "###}",
        },
    });
    // Add filters
    filters.forEach((filterFn, filterName) => {
        env.addFilter(filterName, filterFn.bind(context));
    });
    // Add blocks
    blocks.forEach((block) => {
        const extName = block.getExtensionName();
        const Ext = block.toNunjucksExt(context, blocksOutput);
        env.addExtension(extName, new Ext());
    });
    // Add globals
    globals.forEach((globalValue, globalName) => {
        env.addGlobal(globalName, globalValue);
    });
    // Add other extensions
    extensions.forEach((ext, extName) => {
        env.addExtension(extName, ext);
    });
    return env;
};
/**
 Create a template engine

 @param {Object} def
 @return {TemplateEngine}
 */
// @ts-expect-error ts-migrate(2339) FIXME: Property 'create' does not exist on type 'Class'.
TemplateEngine.create = function (def) {
    return new TemplateEngine({
        blocks: immutable_1.default.List(def.blocks || []),
        extensions: immutable_1.default.Map(def.extensions || {}),
        filters: immutable_1.default.Map(def.filters || {}),
        globals: immutable_1.default.Map(def.globals || {}),
        context: def.context,
        loader: def.loader,
    });
};
exports.default = TemplateEngine;
